src/node/miner.cpp

```diff
@@ -394,9 +395,7 @@ void BlockAssembler::addPackageTxs(const CTxMemPool& mempool, int& nPackagesSele
             continue;
         }
 
-        CTxMemPool::setEntries ancestors;
-        std::string dummy;
-        mempool.CalculateMemPoolAncestors(*iter, ancestors, CTxMemPool::Limits::NoLimits(), dummy, false);
+        auto ancestors{*Assert(mempool.CalculateMemPoolAncestors(*iter, CTxMemPool::Limits::NoLimits(), /*fSearchForParents=*/false))};
```

Comment:
Even when this currently cannot fail (because of the no limits and search parents args), I would be more conservative and catch + log any possible result error instead of assert the result.

In this way, we keep the current behavior, and use an empty `ancestors` set instead of aborting in production if something gets changed inside `CalculateMemPoolAncestors` and we forget to update this call.

Reply:
Interesting. I see your point re robustness, but generally, I really don't like silent failures - even when logged. I'd much rather they throw so we can more easily detect and fix unexpected behaviour as early on as possible. There are 7 instances in this PR where we do this kind of assertion - would you only want to update it here or also in the other instances? I also understand that the current behaviour is that of silent failure, so I wouldn't want to break anything with this change - but as you mention, I don't think it does, given the `NoLimits()` parameter in all of these instances.

Cc'ing the other reviewers in case anyone else has a view (@w0xlt @glozow @theStack @aureleoules)

<details>
<summary>git diff</summary>

```diff
diff --git a/src/node/miner.cpp b/src/node/miner.cpp
index 0dc50c75c..c6838aee0 100644
--- a/src/node/miner.cpp
+++ b/src/node/miner.cpp
@@ -395,7 +395,13 @@ void BlockAssembler::addPackageTxs(const CTxMemPool& mempool, int& nPackagesSele
             continue;
         }
 
-        auto ancestors{*Assert(mempool.CalculateMemPoolAncestors(*iter, CTxMemPool::Limits::NoLimits(), /*fSearchForParents=*/false))};
+        auto ancestors_result{mempool.CalculateMemPoolAncestors(*iter, CTxMemPool::Limits::NoLimits(), /*fSearchForParents=*/false)};
+        if (!ancestors_result) {
+            LogPrintLevel(BCLog::MEMPOOL, BCLog::Level::Error, 
+                          "%s: CalculateMemPoolAncestors failed unexpectedly, continuing with empty ancestor set (%s)",
+                          __func__, util::ErrorString(ancestors_result));
+        }
+        auto ancestors{ancestors_result.value_or(CTxMemPool::setEntries{})}; // for robustness, continue with empty ancestors if the calculation failed
 
         onlyUnconfirmed(ancestors);
         ancestors.insert(iter);
```
</details>

Reply:
At least you'll still need to use `Assume` to abort in debug builds

Reply:
> Interesting. I see your point re robustness, but generally, I really don't like silent failures - even when logged. I'd much rather they throw so we can more easily detect and fix unexpected behaviour as early on as possible

Well, at least on the mempool and validation areas, I'm on the totally opposite position.
If a result can have different meanings, then I would forcibly handle every possible outcome. We want robustness there.

Generally speaking: as the node's mempool state is built from the network, if there is a way to trigger something like this from an external source and we overseeing it (which could happen because we are humans), then an attacker could crash nodes remotely. And once that happens, there isn't a quick fix. Core isn't a common piece of software, any new version distribution and update coordination is hard.

Thus why I would rather be more conservative in terms of using assertions there. And instead, I would actually force us to not ignore errors paths (we could extend our result class to work similarly to the [Rust result type](https://doc.rust-lang.org/std/result/), so we are always forced to handle every possible outcome by default) or to rewrite code so it cannot fail in places that we don't expect failures.

This all comes from very bad experiences that I had in the past..

> At least you'll still need to use Assume to abort in debug builds

@MarcoFalke that wouldn't work, the result class internally asserts that the value exists when the `operator*()` is called (the upper `Assert()` call is actually redundant).

Reply:
> @MarcoFalke that wouldn't work, the result class internally asserts that the value exists when the operator*() is called (the upper Assert() call is actually redundant).

Sure, but not when `*` is replaced with `value_or`.

Reply:
> > @MarcoFalke that wouldn't work, the result class internally asserts that the value exists when the operator*() is called (the upper Assert() call is actually redundant).
> 
> Sure, but not when `*` is replaced with `value_or`.

Yeah, in that way we would retain the current behavior and add the assertion for debug builds only.

Still, future thing, I'm more inclined to rewrite the code so the function cannot fail or.., at least, add the error handler so it becomes clear what we do if it fails.

Reply:
> (the upper Assert() call is actually redundant)

I don't think that's true, Assert does a [boolean evaluation](https://github.com/bitcoin/bitcoin/blob/master/src/util/check.h#L66), so Result's `operator*()` [is not called](https://github.com/bitcoin/bitcoin/blob/3db23fd821e0a08e246a60c198676be42de5bcd6/src/util/result.h#L70).

> ... then an attacker could crash nodes remotely...

Point taken. I've reorganized (force pushing soon) the PR so that the first 2 commits keep the current (robust) behaviour but just using `util::Result`. In 2 subsequent commits I introduce and use a helper fn `AssumeCalculateMemPoolAncestors` to call `CalculateMemPoolAncestors` and always get a setEntries back, but for non-debug builds log the error in case of failure, or halt the application for debug builds. I think that strikes a nice balance between robustness in production and increasing visibility (of which currently we have 0 in master, for this issue) in development - hope you agree?

Reply:
Using `value_or` and `Assume()` seems like a good balance between handling every case and avoiding asserts in p2p-reachable codepaths. It's not great that some callsites ignore the return value of CMPA and continue using the ancestor set on master - I was going to comment about it until I saw the last couple commits. Agree adding an `Assume()` to catch in debug mode + logging is a better approach than halting in production.

Reply:
I think everyone's concerns are now addressed so I'll mark this as resolved, thanks for the input everyone.